allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,244 @@
+                                                     flat_head_idx).squeeze(-1) # [b, num_spans, max_span_width]
+        span_head_scores += head_mask.float().log() # [b, num_spans, max_span_width]
+        flat_span_head_scores = span_head_scores.view(-1, self._max_span_width) # [b * num_spans, max_span_width]
+        flat_span_head_weights = F.softmax(flat_span_head_scores) # [b * num_spans, max_span_width]
+        flat_span_head_weights = flat_span_head_weights.unsqueeze(1) # [b * num_spans, 1, max_span_width]
+        flat_span_text_emb = span_text_emb.view(-1,
+                                                self._max_span_width,
+                                                span_text_emb.size(-1)) # [b * num_spans, max_span_width, emb]
+        flat_attended_text_emb = flat_span_head_weights.bmm(flat_span_text_emb) # [b * num_spans, 1, emb]
+        attended_text_emb = flat_attended_text_emb.view(batch_size, num_spans, -1) # [b, num_spans, emb]
+        span_emb = torch.cat([start_emb, end_emb, span_size_emb, attended_text_emb], -1) # [b, num_spans, emb]
+
+        # Compute mention scores.
+        mention_scores = self._mention_scorer(self._mention_feedforward(span_emb)) # [b, num_spans, 1]
+        mention_scores += span_mask.log() # [b, num_spans, 1]
+
+        # Prune based on mention scores.
+        k = int(math.floor(self._spans_per_word * text_len)) # []
+        _, top_span_idx = mention_scores.topk(k, 1) # [b, k, 1]
+        top_span_idx, _ = torch.sort(top_span_idx, 1) # [b, k, 1]
+        top_span_idx = top_span_idx.squeeze(-1) # [b, k]
+        flat_top_span_idx = util.flatten_batched_indices(top_span_idx, num_spans) # [b * k]
+
+        # Select tensors relating to the top spans.
+        top_span_mask = util.batched_index_select(span_mask, top_span_idx, flat_top_span_idx) # [b, k, 1]
+        top_span_mention_scores = util.batched_index_select(mention_scores,
+                                                            top_span_idx,
+                                                            flat_top_span_idx) # [b, k, 1]
+        top_span_starts = util.batched_index_select(span_starts, top_span_idx, flat_top_span_idx) # [b, k, 1]
+        top_span_ends = util.batched_index_select(span_ends, top_span_idx, flat_top_span_idx) # [b, k, 1]
+        top_span_emb = util.batched_index_select(span_emb, top_span_idx, flat_top_span_idx) # [b, k, emb]
+        top_span_labels = util.batched_index_select(span_labels.unsqueeze(-1),
+                                                    top_span_idx,
+                                                    flat_top_span_idx) # [b, k, 1]
+
+        # Compute indices for antecedent spans to consider.
+        max_ant = min(self._max_antecedents, k) # []
+        target_idx = util.get_indices(k, is_cuda).unsqueeze(1) # [k, 1]
+        ant_offsets = (util.get_indices(max_ant, is_cuda) + 1).unsqueeze(0) # [1, max_ant]
+        raw_ant_idx = target_idx - ant_offsets # [k, max_ant]
```

Comment:
This is another place where some comments would be particularly helpful.  You're essentially creating a triangular matrix like:

```
0 0 0 0 ...
1 0 0 0 ...
1 2 0 0 ...
1 2 3 0 ...
...
```

Right?

Since you do this in a few places, it's probably worth just making this a two-argument function in `nn.util`, with width and height as arguments (and probably whether the diagonal itself should be included).  And if that function has a good name, no comments needed here.